libs/gl-client/src/node/mod.rs

```diff
@@ -23,6 +25,86 @@ pub trait GrpcClient {
     fn new_with_inner(inner: service::AuthService) -> Self;
 }
 
+/// A builder to set up and configure a `Node`.
+pub struct NodeBuilder {
+    node_id: Vec<u8>,
+    network: Network,
+    tls: Option<TlsConfig>,
+    rune: Option<futhark::Rune>,
+    auth: Option<Vec<u8>>,
+}
+
+impl NodeBuilder {
+    /// Sets a `TlsConfig` for the `Node`. Overrides a `TlsConfig` created from
+    /// the auth blob using `with_auth`.
+    pub fn with_tls(self, tls: TlsConfig) -> Self {
+        Self {
+            node_id: self.node_id,
+            network: self.network,
+            tls: Some(tls),
+            rune: self.rune,
+            auth: self.auth,
+        }
+    }
+
+    /// Sets a `futhark::Rune` for the `Node`. Overrides a `futhark::Rune` that
+    /// is extracted from the auth blob via `with_auth`.
+    pub fn with_rune(self, rune: futhark::Rune) -> Self {
+        Self {
+            node_id: self.node_id,
+            network: self.network,
+            tls: self.tls,
+            rune: Some(rune),
+            auth: self.auth,
+        }
+    }
+
+    /// Use the auth blob to create the `TlsConfig` and the `futhark::Rune` for
+    /// the `Node`. Will be overridden by `with_rune()` and `with_tls`.
+    pub fn with_auth(self, auth: &[u8]) -> Self {
+        Self {
+            node_id: self.node_id,
+            network: self.network,
+            tls: self.tls,
+            rune: self.rune,
+            auth: Some(auth.to_vec()),
+        }
+    }
+
+    /// Build the `Node` from set parameters.
+    pub fn build(self) -> Result<Node> {
+        let mut tls = None;
+        let mut rune = None;
+
+        if let Some(auth) = self.auth {
+            // deserialize auth blob and create identity;
+            let blob = serialize::AuthBlob::deserialize(&auth[..])?;
+            tls = Some(TlsConfig::new()?.identity(blob.cert, blob.key));
+            let srune = Rune::from_base64(&blob.rune)?;
+            rune = Some(srune);
+        }
+
+        if let Some(stls) = self.tls {
+            tls = Some(stls);
+        }
+
+        if let Some(srune) = self.rune {
+            rune = Some(srune);
+        }
```

Comment:
Mh, actually they are not really, they add order.
1. First we check if `auth` is set and if so, we deserialize the blob and create the `TlsConfig` and the `Rune` and store them in the variables we initialized with `None`.

2. The second step now is to override the `TlsConfig` and the `Rune` if, and only if they have been set via `with_...`. Otherwise we would override the values from the auth blob with `None`.

But looking at this again it seems off. I should probably remove this altogether and just operate on `self` just as you suggested. The order of invoking `with_...` is responsible then for overriding previous set fields.